wal/wal.go

```diff
@@ -707,25 +706,38 @@ func (r *segmentBufReader) Close() (err error) {
 	return err
 }
 
+// Read implements io.Reader, expect that we occasionally return err=nil
+// when n != len(b), which violates the inteface.  Only use this with
+// io.ReadFull.
 func (r *segmentBufReader) Read(b []byte) (n int, err error) {
-	if !r.more {
-		if r.cur+1 >= len(r.segs) {
-			return 0, io.EOF
-		}
-		r.cur++
-		r.off = 0
-		r.more = true
-		r.buf.Reset(r.segs[r.cur])
-	}
 	n, err = r.buf.Read(b)
 	r.off += n
+
+	// If we succeeded, or hit a non-EOF, we can stop.
 	if err != io.EOF {
 		return n, err
 	}
-	// Just return what we read so far, but don't signal EOF.
-	// Only unset more so we don't invalidate the current segment and
-	// offset before the next read.
-	r.more = false
+
+	// We hit EOF; fake out zero padding at the end of short segments.
```

Comment:
This is the whole point of the PR.   From the description:

> The test fails against master, as the underlying segmentReader has already moved on to the next segment.

> make the segmentReader pad short segments with zeros, and only advance to next segment on the call to `Read` _after_ completely consuming this one

> previously it would increment `curr` on the call to `Read` which failed, causing the returns `CorruptionErr` to reference the wrong segment.

Reply:
Yeah sorry I meant to add some more details in the comment.

Reply:
Done!